hw/ip/otbn/data/otbn.hjson

```diff
@@ -102,30 +102,79 @@
         }
       ]
     } // register : status
-    { name: "ERR_CODE",
-      desc: "Error Code",
+    { name: "ERR_BITS",
+      desc: '''
+        Error bitfield. Reads as non-zero if an error was seen during OTBN
+        operation
+      ''',
       swaccess: "ro",
       hwaccess: "hwo",
       fields: [
-        { bits: "31:0",
-          name: "err_code",
+        { bits: "0",
+          name: "bad_data_addr"
           desc: '''
-            The error cause if an error occurred.
-
-            This is set to zero (ErrCodeNoError) when OTBN successfully
-            completes an operation or a positive value when OTBN stops on an
-            error.
-
-            While OTBN is running (!!CMD.busy set), this has no meaning. When
-            OTBN stops this will be set to indicate the outcome. It is
-            guaranteed not to change until OTBN is run again (with !!CMD.start).
-
-            See <a href='#error-conditions'>Error Conditions</a> for
-            details on the possible values.
+            A DMEM read or write occurred with an out of bounds or unaligned
+            address.
           '''
         }
+        { bits: "1",
+          name: "bad_insn_addr"
+          desc: '''
+            An IMEM read or write occurred with an out of bounds or unaligned
+            address.
+          '''
+        }
+        { bits: "2",
+          name: "call_stack"
+          desc: '''
+            A instruction tried to pop from an empty call stack or push to a
+            full call stack.
+          '''
+        }
+        { bits: "3",
+          name: "illegal_insn"
+          desc: '''
+            One of the following happened:
+            <ul>
+              <li>An instruction being excuted had an invalid encoding.</li>
+              <li>An access occurred for an invalid CSR or WSR.</li>
+              <li>
+                A CSR or WSR access occurred that is not permitted (e.g. writing
+                to a read-only CSR or WSR).
+              </li>
+            </ul>
+          '''
+        }
+        { bits: "4",
+          name: "loop"
+          desc: '''
+            One of the following happened:
+            <ul>
+              <li>A loop was started with an iteration count of zero.</li>
+              <li>
+                The final instruction of a loop was a branch or another loop.
+              </li>
+              <li>
+                A new loop tried to push to a full loop stack (loop nesting
+                level too deep).
+              </li>
+            </ul>
+          '''
+        }
+        { bits: "5",
+          name: "fatal_imem"
```

Comment:
First thought: To keep fatal and non-fatal errors grouped and avoid some reordering in the future, could we put those fatal errors into the top of the bitfield (bits 31 and down)?

Second thought: This would make the status code passing from simulation more annoying (i.e. break our current "hack"), so maybe it's not worth it.

Reply:
We could always alter the model wrapper so it passes error out in a seperate 32-bit field, or something nastier like take the bits from the model and split them out into the right places (but then ISS and RTL have different bit positions for the same error)

Reply:
That modification to the wrapper might be a good idea either way: packing bits in a uint32 always felt like a bit of a hack. I'll try doing that now and see how it looks (it's sort of orthogonal from this change, although there might be a few conflicting lines).

Reply:
Oops, I forgot to link back. This re-packing appears as #4850. I'm not sure what order you think it's better to merge things.

Reply:
I don't think it really matters, should be a pretty trivial conflict to fix in either PR, lets just see who hits merge first!

Reply:
Now with #4850 in, can we push the fatal alert bits to the top?

Reply:
We can do though it once again raises a question of how we deal with that from the model's perspective.

Both RTL and model wrapper produce an instance of `err_bits_t` which would not include the gap (it's just a list of named bits). The gap comes in when those individual fields are assigned out to the ERR_CODE/ERR_BITS register.

So if I push things to the top I'd need to leave the bit constants in the ISS as they are so things still match up with the internal bit order in the struct. 

I could dump the struct, switch to a 32-bit logic everywhere instead and use parameters (or an enum) to name the bits so I'd be doing: ```assign err_bits_o[ErrBitFatalDmem] = dmem_alert``` in the RTL.

I could also add an `unused` field to the struct to make its internal bit layout match the ERR_CODE/ERR_BITS layout.

@rswarbrick what do you think?

Reply:
I think the best thing is probably to keep the current packed layout, but update the `set_err_code` in `otbn_model.cc` to match the layout of the register.

Reply:
With the current setup if we did that we'd output bits matching the register layout as err_code and then immediately repack them into an `err_bits_t` structure. We'd also add a new place we need to adjust things when the error bits layout change.

Ultimately it's awkward as within otbn_core the bit layout in the register is irrelevant, it simply has a bunch of named errors it pushes out as a struct, which is then fed into a register with a specific layout. The model operates at the same level as otbn_core so needs to produce the same bundle of named errors.

Having the model just produce the bundle of named errors could be sensible, though there's still the question of how that crosses the python <-> c <-> systemverilog interface, plus falls into the category of changes we were aiming to avoid in this PR.

Maybe I should just leave things as is for the this PR (so in particular no moving the fatal alert bits of the top) and come back to this question?

Reply:
Yeah, let's come back to that in a follow-up.